Signal/src/contact/SystemContactsFetcher.swift

```diff
@@ -0,0 +1,120 @@
+}
+
+@objc
+class SystemContactsFetcher: NSObject {
+
+    private let TAG = "[SystemContactsFetcher]"
+
+    public weak var delegate: SystemContactsFetcherDelegate?
+
+    public var authorizationStatus: CNAuthorizationStatus {
+        return CNContactStore.authorizationStatus(for: CNEntityType.contacts)
+    }
+
+    public var isAuthorized: Bool {
+        guard self.authorizationStatus != .notDetermined else {
+            assertionFailure("should have called `requestOnce` before this point.")
+            Logger.error("\(TAG) should have called `requestOnce` before checking authorization status.")
+            return false
+        }
+
+        return self.authorizationStatus == .authorized
+    }
+
+    private let contactStore = CNContactStore()
+    private var systemContactsHaveBeenRequestedAtLeastOnce = false
+    private let allowedContactKeys: [CNKeyDescriptor] = [
+        CNContactFormatter.descriptorForRequiredKeys(for: .fullName),
+        CNContactThumbnailImageDataKey as CNKeyDescriptor, // TODO full image instead of thumbnail?
+        CNContactPhoneNumbersKey as CNKeyDescriptor,
+        CNContactEmailAddressesKey as CNKeyDescriptor
+    ]
+
+    public func requestOnce() {
+        AssertIsOnMainThread()
+
+        guard !systemContactsHaveBeenRequestedAtLeastOnce else {
+            Logger.debug("\(TAG) already requested system contacts")
+            return
+        }
+        systemContactsHaveBeenRequestedAtLeastOnce = true
+        self.startObservingContactChanges()
+
+        switch authorizationStatus {
+        case .notDetermined:
+            contactStore.requestAccess(for: .contacts, completionHandler: { (granted, error) in
+                if let error = error {
+                    Logger.error("\(self.TAG) error fetching contacts: \(error)")
+                    assertionFailure()
+                }
+
+                if !granted {
+                    // TODO, make this a one time dismissable admonishment
+                    // e.g. remember across launches that the user has dismissed.
+                    self.displayMissingContactsPermissionAlert()
+                } else {
+                    self.updateContacts()
+                }
+            })
+        case .authorized:
+            // TODO reset onetime admonishment reminder, so that we remind user again (once) if they've since toggled permissions.
+            self.updateContacts()
+        case .denied, .restricted:
+            Logger.debug("\(TAG) contacts were \(self.authorizationStatus)")
+        }
+    }
+
+    public func fetchIfAlreadyAuthorized() {
+        AssertIsOnMainThread()
+        guard authorizationStatus == .authorized else {
+            return
+        }
+
+        updateContacts()
+    }
+
+    private func displayMissingContactsPermissionAlert() {
+        let foo = UIApplication.shared.frontmostViewController
+        Logger.error("TODO")
```

Comment:
This is kind of a design TODO. I should have documented my thoughts rather than leaving this TODO. I don't think we actually want to risk enraging hitting the user with a "shame on you" alert immediately after they say "no" to the previous alert. 

I think we will be better served by the contextually relevant reminders (e.g. in the contact list, something like "Your contact list is empty because you haven't enabled contacts access. [Settings]")

WDYT?